pkg/store/bucket.go

```diff
@@ -1352,150 +1405,139 @@ func (r *bucketIndexReader) ExpandedPostings(ms []*labels.Matcher) ([]uint64, er
 	return ps, nil
 }
 
-type postingGroup struct {
-	keys     labels.Labels
-	postings []index.Postings
+func getAllPostingsKeyLabel() labels.Label {
+	name, value := index.AllPostingsKey()
+	return labels.Label{Name: name, Value: value}
+}
 
-	aggregate func(postings []index.Postings) index.Postings
+// Logical result of each individual group is:
+// If addAll is set: ALL postings minus postings for removeKeys labels. (No need to merge postings for addKeys in this case)
+// If addAll is not set: Merge of postings for "addKeys" labels minus postings for removeKeys labels
+// This happens in ExpandedPostings.
+type postingGroup struct {
+	addAll     bool
+	addKeys    []labels.Label
+	removeKeys []labels.Label
 }
 
-func newPostingGroup(keys labels.Labels, aggr func(postings []index.Postings) index.Postings) *postingGroup {
+func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) *postingGroup {
 	return &postingGroup{
-		keys:      keys,
-		postings:  make([]index.Postings, len(keys)),
-		aggregate: aggr,
+		addAll:     addAll,
+		addKeys:    addKeys,
+		removeKeys: removeKeys,
 	}
 }
 
-func (p *postingGroup) Fill(i int, posting index.Postings) {
-	p.postings[i] = posting
+// returns true, if this postingGroup will always return empty postings.
+func (p *postingGroup) alwaysEmptyPostings() bool {
+	return !p.addAll && len(p.addKeys) == 0
 }
 
-func (p *postingGroup) Postings() index.Postings {
-	if len(p.keys) == 0 {
-		return index.EmptyPostings()
+func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
+	if p == nil {
+		// This should not happen. Debug for https://github.com/thanos-io/thanos/issues/874.
+		return index.ErrPostings(errors.Errorf("postings is nil for %s. It was never fetched.", l))
 	}
-
-	for i, posting := range p.postings {
-		if posting == nil {
-			// This should not happen. Debug for https://github.com/thanos-io/thanos/issues/874.
-			return index.ErrPostings(errors.Errorf("at least one of %d postings is nil for %s. It was never fetched.", i, p.keys[i]))
-		}
-	}
-
-	return p.aggregate(p.postings)
-}
-
-func merge(p []index.Postings) index.Postings {
-	return index.Merge(p...)
-}
-
-func allWithout(p []index.Postings) index.Postings {
-	return index.Without(p[0], index.Merge(p[1:]...))
+	return p
 }
 
 // NOTE: Derived from tsdb.postingsForMatcher. index.Merge is equivalent to map duplication.
 func toPostingGroup(lvalsFn func(name string) ([]string, error), m *labels.Matcher) (*postingGroup, error) {
-	var matchingLabels labels.Labels
+	// This matches all values, including no value. If it is the only matcher, it will return all postings.
+	if m.Type == labels.MatchRegexp && (m.Value == ".*" || m.Value == "^.*$") {
+		return newPostingGroup(true, nil, nil), nil
+	}
```

Comment:
Why can't we just `return nil, nil`  and then handle the `nil` posting group in `ExpandedPostings()` so that it's not added to list of `postingGroups`? Given this regex match all series, we could safely remove it. If this label matcher is the only one in the query, we could handle that case in `ExpandedPostings()` and return an error (ie. Prometheus doesn't allow to execute such query).

I guess I'm missing obvious legit use cases, but I would like to double check if this logic could be simplified.

Reply:
As discussed privately, we cannot remove this completely. On master, ExpandedPostings basically returns "ALL" postings for single `i=~".*"` matcher. Removing it completely would change the result.